### PR TITLE
OpenAI DALL·E support

### PR DESCRIPTION
**Pull request by Heezer**
Motivating generated picture 😄 ❤️ 
image

### CODE REVIEW

File: langchain4j-open-ai-dall-e/src/main/java/dev/langchain4j/image/openai/dalle/ResponseLoggingInterceptor.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    return response;
+  }
+
+  static void log(Response response) {
+    try {
+      log.debug(
+        "Response:\n- status code: {}\n- headers: {}\n- body: {}",
+        response.code(),
+        RequestLoggingInterceptor.inOneLine(response.headers()),
+        getBody(response)
```
Comment (kumlien) [inline langchain4j-open-ai-dall-e/src/main/java/dev/langchain4j/image/openai/dalle/ResponseLoggingInterceptor.java:29]: Do you want to log the body?

File: langchain4j-open-ai-dall-e/src/main/java/dev/langchain4j/image/openai/dalle/RequestLoggingInterceptor.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+  }
+
+  private static void log(Request request) {
+    try {
+      log.debug(
+        "Request:\n- method: {}\n- url: {}\n- headers: {}\n- body: {}",
+        request.method(),
+        request.url(),
+        inOneLine(request.headers()),
+        getBody(request)
```
Comment (kumlien) [inline langchain4j-open-ai-dall-e/src/main/java/dev/langchain4j/image/openai/dalle/RequestLoggingInterceptor.java:36]: Do you want to log the body?

File: langchain4j-core/src/main/java/dev/langchain4j/data/image/Image.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+    public void setUrl(String url) {
+        this.url = url;
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    public static class Builder {
```
Comment (langchain4j) [inline langchain4j-core/src/main/java/dev/langchain4j/data/image/Image.java:67]: I would replace this builder with a public ctor and `Image.from(String url)` static factory, builder is not very useful for a single field

File: langchain4j-core/src/main/java/dev/langchain4j/data/image/Image.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
+package dev.langchain4j.data.image;
+
+public class Image {
+    private String url;
+
+    public String getUrl() {
```
Comment (langchain4j) [inline langchain4j-core/src/main/java/dev/langchain4j/data/image/Image.java]: minor: please avoid using "get" and "set" in getters and setters, we are following a modern "record" style

File: langchain4j-core/src/main/java/dev/langchain4j/model/chat/ImageProcessingModel.java [hunk 5]
```diff
@@ -0,0 +1,8 @@
+package dev.langchain4j.model.chat;
+
+import dev.langchain4j.data.image.Image;
+import dev.langchain4j.model.output.Response;
+
+public interface ImageProcessingModel {
```
Comment (langchain4j) [inline langchain4j-core/src/main/java/dev/langchain4j/model/chat/ImageProcessingModel.java]: What about renaming it into `ImageGeneratingModel`, as it basically generates an image from a prompt at this point? Processing sounds like it processes images, which is not true, at least at the moment.
DALLE2 and probably other providers also allow to edit and create variations of existing image, so maybe these methods will be added in the future, then we can name it simply `ImageModel`, (as opposed to `LanguageModel`, `EmbeddingModel`, etc). What do you think?

I would also move it into a separate "image" package, as it is not related to a chat.

File: langchain4j-open-ai-dall-e/src/main/java/dev/langchain4j/image/openai/dalle/AuthorizationHeaderInjector.java [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+import org.jetbrains.annotations.NotNull;
+
+@RequiredArgsConstructor
+class AuthorizationHeaderInjector implements Interceptor {
+
+  private final String apiKey;
+
+  @NotNull
+  public Response intercept(Interceptor.Chain chain) throws IOException {
+    Request request = chain.request().newBuilder().addHeader("Authorization", "Bearer " + this.apiKey).build();
```
Comment (langchain4j) [inline langchain4j-open-ai-dall-e/src/main/java/dev/langchain4j/image/openai/dalle/AuthorizationHeaderInjector.java]: All this logic is already implemented in openai4j library, I would suggest moving missing relevant parts to image generation there

File: langchain4j-open-ai-dall-e/pom.xml [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    <modelVersion>4.0.0</modelVersion>
+
+    <parent>
+        <groupId>dev.langchain4j</groupId>
+        <artifactId>langchain4j-parent</artifactId>
+        <version>0.24.0</version>
+        <relativePath>../langchain4j-parent/pom.xml</relativePath>
+    </parent>
+
+    <artifactId>langchain4j-open-ai-dall-e</artifactId>
```
Comment (langchain4j) [inline langchain4j-open-ai-dall-e/pom.xml]: I am not sure we should introduce a new module for a specific OpenAI model, we have all their models already in langchain4j-open-ai

File: langchain4j-open-ai-dall-e/src/main/java/dev/langchain4j/image/openai/dalle/OpenAiDalleClient.java [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+    Duration connectTimeout,
+    Duration readTimeout,
+    Duration writeTimeout,
+    Proxy proxy,
+    boolean logRequests,
+    boolean logResponses
+  ) {
+    OkHttpClient.Builder cBuilder =
+      (new OkHttpClient.Builder()).callTimeout(getOrDefault(callTimeout, ofSeconds(60)))
+        .connectTimeout(getOrDefault(connectTimeout, ofSeconds(60)))
```
Comment (langchain4j) [inline langchain4j-open-ai-dall-e/src/main/java/dev/langchain4j/image/openai/dalle/OpenAiDalleClient.java]: a lot of logic from this class already exists in openai4j

File: pom.xml [hunk 9]
```diff
@@ -28,14 +28,17 @@
...
+
         <!-- embedding stores -->
         <module>langchain4j-cassandra</module>
         <module>langchain4j-chroma</module>
         <module>langchain4j-elasticsearch</module>
         <module>langchain4j-milvus</module>
         <module>langchain4j-opensearch</module>
-        <module>langchain4j-pgvector</module>
         <module>langchain4j-pinecone</module>
+        <module>langchain4j-pgvector</module>
```
Comment (langchain4j) [inline pom.xml]: g comes before i, or? :)

File: langchain4j-open-ai-dall-e/src/main/java/dev/langchain4j/image/openai/dalle/OpenAiDalleImageProcessingModel.java [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+  /**
+   * Instantiates OpenAI DALL·E image processing model.
+   * Find the parameters description <a href="https://platform.openai.com/docs/api-reference/images/create">here</a>.
+   * Only one image can be generated pro request currently.
+   *
+   * @param model dall-e-3 is default one
+   * @param downloadFolder specifies the local folder where the generated image will be downloaded to (in case provided).
+   */
+  @Builder
+  public OpenAiDalleImageProcessingModel(
```
Comment (langchain4j) [inline langchain4j-open-ai-dall-e/src/main/java/dev/langchain4j/image/openai/dalle/OpenAiDalleImageProcessingModel.java]: Maybe OpenAiImageModel? Dalle is a name of the current models, and it is set as a parameter. I can imagine they adding a new generation of models that will be called differently but will have the same API

File: langchain4j-open-ai-dall-e/src/main/java/dev/langchain4j/image/openai/dalle/OpenAiDalleRequest.java [hunk 11]
```diff
@@ -0,0 +1,19 @@
+package dev.langchain4j.image.openai.dalle;
+
+import lombok.Builder;
+import lombok.Data;
+
+@Builder
+@Data
+public class OpenAiDalleRequest {
```
Comment (langchain4j) [inline langchain4j-open-ai-dall-e/src/main/java/dev/langchain4j/image/openai/dalle/OpenAiDalleRequest.java]: Maybe `OpenAiImageGenerationRequest` ?

File: langchain4j-open-ai-dall-e/src/main/java/dev/langchain4j/image/openai/dalle/OpenAiImageDownloader.java [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  private static String extractFileName(String url) {
+    Pattern pattern = Pattern.compile(".*/([^/?]+)\\?.*");
+    Matcher matcher = pattern.matcher(url);
+
+    return matcher.find() ? matcher.group(1) : "image.png";
```
Comment (langchain4j) [inline langchain4j-open-ai-dall-e/src/main/java/dev/langchain4j/image/openai/dalle/OpenAiImageDownloader.java]: for "image.png" case: should we maybe add a timestamp to it?

File: langchain4j-open-ai-dall-e/src/main/java/dev/langchain4j/image/openai/dalle/OpenAiDalleImageProcessingModel.java [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+      .size(size)
+      .model(model)
+      .quality(quality)
+      .style(style)
+      .build();
+
+    OpenAiDalleResponse response = withRetry(() -> client.generate(request), maxRetries);
+    String url = response.getData().get(0).getUrl();
+    if (downloadFolder != null) {
+      url = OpenAiImageDownloader.downloadFile(url, downloadFolder);
```
Comment (langchain4j) [inline langchain4j-open-ai-dall-e/src/main/java/dev/langchain4j/image/openai/dalle/OpenAiDalleImageProcessingModel.java]: what about doing downloading also withRetry()?

File: langchain4j-core/src/main/java/dev/langchain4j/data/image/Image.java [hunk 14]
```diff
@@ -0,0 +1,35 @@
+package dev.langchain4j.data.image;
+
+public class Image {
+    private String url;
```
Comment (langchain4j) [inline langchain4j-core/src/main/java/dev/langchain4j/data/image/Image.java]: Not sure if it is a good idea, but what about using `URI` instead of a `String` here?

File: langchain4j-open-ai-dall-e/src/main/java/dev/langchain4j/image/openai/dalle/OpenAiDalleImageProcessingModel.java [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+  public static final String DALL_E_QUALITY_STANDARD = "standard";
+  public static final String DALL_E_QUALITY_HD = "hd";
+  public static final String DALL_E_STYLE_VIVID = "vivid";
+  public static final String DALL_E_STYLE_NATURAL = "natural";
+
+  private final String model;
+  private final String size;
+  private final String quality;
+  private final String style;
+  private final String downloadFolder;
```
Comment (langchain4j) [inline langchain4j-open-ai-dall-e/src/main/java/dev/langchain4j/image/openai/dalle/OpenAiDalleImageProcessingModel.java]: nitpick: what about `String downloadFolder` -> `Path downloadDirectory` ?

File: langchain4j-open-ai-dall-e/src/main/java/dev/langchain4j/image/openai/dalle/OpenAiDalleResponse.java [hunk 16]
```diff
@@ -0,0 +1,17 @@
+package dev.langchain4j.image.openai.dalle;
+
+import java.util.List;
+import lombok.Data;
+
+@Data
+class OpenAiDalleResponse {
```
Comment (langchain4j) [inline langchain4j-open-ai-dall-e/src/main/java/dev/langchain4j/image/openai/dalle/OpenAiDalleResponse.java]: Maybe `OpenAiImageGenerationResponse`?

**Review by ghost — COMMENTED** [review]
@Heezer thanks a lot! Please check my comments.
I trust that it works, but I would add at least one integration test to make sure it keeps working in the future.
Thanks!

File: langchain4j-core/src/main/java/dev/langchain4j/data/image/Image.java (hunk 3 above)
Comment (Heezer) [reply langchain4j-core/src/main/java/dev/langchain4j/data/image/Image.java:67]: I've just implemented the initial simplified version here with `url` field only, the next would be `b64_json` and `revised_prompt`, see https://platform.openai.com/docs/api-reference/images and it still makes sense from my side to keep this class generic, therefore I've put it into core. Tell me if you still see the changes needed here.

File: langchain4j-core/src/main/java/dev/langchain4j/data/image/Image.java (hunk 4 above)
Comment (Heezer) [reply langchain4j-core/src/main/java/dev/langchain4j/data/image/Image.java]: done

File: langchain4j-core/src/main/java/dev/langchain4j/model/chat/ImageProcessingModel.java (hunk 5 above)
Comment (Heezer) [reply langchain4j-core/src/main/java/dev/langchain4j/model/chat/ImageProcessingModel.java]: Yes, my intention about the name was to be more generic in case of https://platform.openai.com/docs/api-reference/images/createEdit, indeed this is supported by DALL·E 2 only. I'm fine with `ImageModel` and to be honest I cannot remember why I've put it into `chat` 😄, `image` package makes more sense 😄

Comment (Heezer) [reply langchain4j-core/src/main/java/dev/langchain4j/model/chat/ImageProcessingModel.java]: done

File: langchain4j-open-ai-dall-e/src/main/java/dev/langchain4j/image/openai/dalle/AuthorizationHeaderInjector.java (hunk 6 above)
Comment (Heezer) [reply langchain4j-open-ai-dall-e/src/main/java/dev/langchain4j/image/openai/dalle/AuthorizationHeaderInjector.java]: yes, auth and logging interceptors are there, but I was not sure who is in charge for `openai4j` and whether I needs to touch it 😄 ok, now I know this lib is yours as well and if you suggest to move it...

File: langchain4j-open-ai-dall-e/pom.xml (hunk 7 above)
Comment (Heezer) [reply langchain4j-open-ai-dall-e/pom.xml]: hm... why not? going to pack it into existing `langchain4j-open-ai` then

File: pom.xml (hunk 9 above)
Comment (Heezer) [reply pom.xml]: who cares? 😂 fixed

File: langchain4j-open-ai-dall-e/src/main/java/dev/langchain4j/image/openai/dalle/AuthorizationHeaderInjector.java (hunk 6 above)
Comment (Heezer) [reply langchain4j-open-ai-dall-e/src/main/java/dev/langchain4j/image/openai/dalle/AuthorizationHeaderInjector.java]: I've finally managed to refactor the main logic to openai4 lib, so please review it: https://github.com/ai-for-java/openai4j/pull/5

**Comment by Heezer** [comment]
> @Heezer thanks a lot! Please check my comments. I trust that it works, but I would add at least one integration test to make sure it keeps working in the future. Thanks!

integra test is within openai4j project

File: langchain4j-open-ai-dall-e/pom.xml (hunk 7 above)
Comment (Heezer) [reply langchain4j-open-ai-dall-e/pom.xml]: done

File: langchain4j-open-ai-dall-e/src/main/java/dev/langchain4j/image/openai/dalle/OpenAiDalleClient.java (hunk 8 above)
Comment (Heezer) [reply langchain4j-open-ai-dall-e/src/main/java/dev/langchain4j/image/openai/dalle/OpenAiDalleClient.java]: refactored

File: langchain4j-open-ai-dall-e/src/main/java/dev/langchain4j/image/openai/dalle/OpenAiDalleImageProcessingModel.java (hunk 10 above)
Comment (Heezer) [reply langchain4j-open-ai-dall-e/src/main/java/dev/langchain4j/image/openai/dalle/OpenAiDalleImageProcessingModel.java]: done

File: langchain4j-open-ai-dall-e/src/main/java/dev/langchain4j/image/openai/dalle/OpenAiDalleRequest.java (hunk 11 above)
Comment (Heezer) [reply langchain4j-open-ai-dall-e/src/main/java/dev/langchain4j/image/openai/dalle/OpenAiDalleRequest.java]: done in openai4j

File: langchain4j-open-ai-dall-e/src/main/java/dev/langchain4j/image/openai/dalle/OpenAiImageDownloader.java (hunk 12 above)
Comment (Heezer) [reply langchain4j-open-ai-dall-e/src/main/java/dev/langchain4j/image/openai/dalle/OpenAiImageDownloader.java]: done in openai4j

File: langchain4j-open-ai-dall-e/src/main/java/dev/langchain4j/image/openai/dalle/OpenAiDalleImageProcessingModel.java (hunk 13 above)
Comment (Heezer) [reply langchain4j-open-ai-dall-e/src/main/java/dev/langchain4j/image/openai/dalle/OpenAiDalleImageProcessingModel.java]: this part resides in openai4j now

File: langchain4j-core/src/main/java/dev/langchain4j/data/image/Image.java (hunk 14 above)
Comment (Heezer) [reply langchain4j-core/src/main/java/dev/langchain4j/data/image/Image.java]: done in openai4j

File: langchain4j-open-ai-dall-e/src/main/java/dev/langchain4j/image/openai/dalle/OpenAiDalleImageProcessingModel.java (hunk 15 above)
Comment (Heezer) [reply langchain4j-open-ai-dall-e/src/main/java/dev/langchain4j/image/openai/dalle/OpenAiDalleImageProcessingModel.java]: done in openai4j

File: langchain4j-open-ai-dall-e/src/main/java/dev/langchain4j/image/openai/dalle/OpenAiDalleResponse.java (hunk 16 above)
Comment (Heezer) [reply langchain4j-open-ai-dall-e/src/main/java/dev/langchain4j/image/openai/dalle/OpenAiDalleResponse.java]: done in openai4j

**Comment by Heezer** [comment]
> @Heezer thanks a lot! Please check my comments. I trust that it works, but I would add at least one integration test to make sure it keeps working in the future. Thanks!

inte test added

File: langchain4j-core/src/main/java/dev/langchain4j/data/image/Image.java [hunk 17]
```diff
@@ -0,0 +1,69 @@
+package dev.langchain4j.data.image;
+
+import java.net.URI;
+import java.util.Objects;
+
+public final class Image {
+
+    private URI url;
+    private String revisedPrompt;
```
Comment (langchain4j) [inline langchain4j-core/src/main/java/dev/langchain4j/data/image/Image.java:12]: I am not sure we should leak this field, at least right now (we can always add it later), as it is OpenAI-specific and it is not clear if this approach will be adopted by other model provider(s)

File: langchain4j-core/src/main/java/dev/langchain4j/data/image/Image.java [hunk 18]
```diff
@@ -0,0 +1,69 @@
+package dev.langchain4j.data.image;
+
+import java.net.URI;
+import java.util.Objects;
+
+public final class Image {
+
+    private URI url;
```
Comment (langchain4j) [inline langchain4j-core/src/main/java/dev/langchain4j/data/image/Image.java:10]: I think we should add a base64 string here as an alternative to URL

File: langchain4j-core/src/main/java/dev/langchain4j/data/image/Image.java [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(url, revisedPrompt);
+    }
+
+    @Override
+    public String toString() {
+        return "Image{" + "url='" + url + '\'' + ", revisedPrompt='" + revisedPrompt + '\'' + '}';
```
Comment (langchain4j) [inline langchain4j-core/src/main/java/dev/langchain4j/data/image/Image.java]: minor: there is `Utils.quoted()`, see `ToolParameters.toString()` as an example

File: langchain4j-core/src/main/java/dev/langchain4j/model/image/ImageModel.java [hunk 20]
```diff
@@ -0,0 +1,9 @@
+package dev.langchain4j.model.image;
+
+import dev.langchain4j.data.image.Image;
+import dev.langchain4j.model.output.Response;
+import java.util.List;
+
+public interface ImageModel {
+  Response<List<Image>> generate(String prompt);
```
Comment (langchain4j) [inline langchain4j-core/src/main/java/dev/langchain4j/model/image/ImageModel.java]: I propose to simplify a bit and return a single image in the API to be consistent with other our model classes that return a single answer for a single prompt. Dalle-3 anyway supports only n=1 and user can always call the model multiple times if he wants multiple images, but I doubt it will be used often in programmatic way as images cost quite a lot. We can always add another method that returns a list in the future (if it will be indeed needed).

File: langchain4j-open-ai/src/main/java/dev/langchain4j/model/openai/OpenAiImageModel.java [hunk 21]
```diff
@@ -0,0 +1,149 @@
...
+            .persistTo(persistTo);
+
+        if (withPersisting != null && withPersisting) {
+            cBuilder.withPersisting();
+        }
+
+        this.client = cBuilder.build();
+
+        this.maxRetries = getOrDefault(maxRetries, 3);
+        this.model = model;
```
Comment (langchain4j) [inline langchain4j-open-ai/src/main/java/dev/langchain4j/model/openai/OpenAiImageModel.java:89]: I would add a default model to be consistent with other model classes and be able to create a model just by providing an api key.

File: langchain4j-open-ai/src/main/java/dev/langchain4j/model/openai/OpenAiImageModel.java [hunk 22]
```diff
@@ -0,0 +1,149 @@
...
+            .builder()
+            .prompt(prompt)
+            .size(size)
+            .quality(quality)
+            .style(style)
+            .user(user)
+            .responseFormat(responseFormat);
+
+        if (DALL_E_2.equals(model)) {
+            requestBuilder.model(dev.ai4j.openai4j.image.ImageModel.DALL_E_2);
```
Comment (langchain4j) [inline langchain4j-open-ai/src/main/java/dev/langchain4j/model/openai/OpenAiImageModel.java:145]: minor: would be nice to have overloaded `model(String)` method in openai4j (as well as change `Builder.model` from `ImageModel` to `String`) to not limit only to these models in enum and not to force to make a conversion from string to enum here

File: langchain4j-open-ai/src/main/java/dev/langchain4j/model/openai/OpenAiImageModel.java [hunk 23]
```diff
@@ -0,0 +1,149 @@
...
+        }
+
+        if (n != null) {
+            requestBuilder.n(n);
```
Comment (langchain4j) [inline langchain4j-open-ai/src/main/java/dev/langchain4j/model/openai/OpenAiImageModel.java]: minor: this is why I use object types instead of primitives in configs, it is easier to pass them through (no need to check if it is null or not) and also there is an understanding if user defined a config or not (primitive will always have a value).

File: langchain4j-open-ai/src/main/java/dev/langchain4j/model/openai/OpenAiImageModel.java [hunk 24]
```diff
@@ -0,0 +1,149 @@
...
+
+        return Response.from(
+            response
+                .data()
+                .stream()
+                .map(d -> Image.builder().url(d.url()).revisedPrompt(d.revisedPrompt()).build())
+                .collect(Collectors.toList())
+        );
+    }
+}
```
Comment (langchain4j) [inline langchain4j-open-ai/src/main/java/dev/langchain4j/model/openai/OpenAiImageModel.java:150]: Could you please add a static `withApiKey(String apiKey)` factory method for parity with other models?

File: langchain4j-open-ai/src/test/java/dev/langchain4j/model/openai/OpenAiImagesGenerationIT.java [hunk 25]
```diff
@@ -0,0 +1,105 @@
...
+import dev.langchain4j.data.image.Image;
+import dev.langchain4j.model.output.Response;
+import java.io.File;
+import java.net.URI;
+import java.util.List;
+import org.junit.jupiter.api.Test;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+class OpenAiImagesGenerationIT {
```
Comment (langchain4j) [inline langchain4j-open-ai/src/test/java/dev/langchain4j/model/openai/OpenAiImagesGenerationIT.java]: minor: `OpenAiImageModelIT`

File: langchain4j-open-ai/src/main/java/dev/langchain4j/model/openai/OpenAiImageModel.java [hunk 26]
```diff
@@ -0,0 +1,149 @@
...
+
+    public static class OpenAiImageModelBuilder {
+
+        public OpenAiImageModelBuilder withPersisting() {
+            withPersisting = true;
+            return this;
+        }
+
+        public OpenAiImageModelBuilder logRequests() {
+            logRequests = true;
```
Comment (langchain4j) [inline langchain4j-open-ai/src/main/java/dev/langchain4j/model/openai/OpenAiImageModel.java]: This is a really nice change but I would propose to make it in a separate PR and update all model classes at once to keep the consistency

**Review by ghost — CHANGES_REQUESTED** [review]
@Heezer great job, thank you! Please check comments

File: langchain4j-core/src/main/java/dev/langchain4j/model/image/ImageModel.java (hunk 20 above)
Comment (langchain4j) [reply langchain4j-core/src/main/java/dev/langchain4j/model/image/ImageModel.java]: @jdubois please check this as well

Comment (jdubois) [reply langchain4j-core/src/main/java/dev/langchain4j/model/image/ImageModel.java]: Yes, I already have this, thanks!

Comment (jdubois) [reply langchain4j-core/src/main/java/dev/langchain4j/model/image/ImageModel.java]: I haven't pushed yet because of the base64 part, it will be there soon

Comment (langchain4j) [reply langchain4j-core/src/main/java/dev/langchain4j/model/image/ImageModel.java]: @jdubois this is to your comment
> This is fixing my 2 biggest issues with this API (not limiting to just 1 image, and being able to get the image content and not just the URL)


I am not convinced we should return a List, I would start from a single Image and later on we can add another mothod to return more (if required)

Comment (jdubois) [reply langchain4j-core/src/main/java/dev/langchain4j/model/image/ImageModel.java]: We could do both....
Also we need to add a new parameter here to support both base64 and URL, I'll give it a try in my PR

Comment (langchain4j) [reply langchain4j-core/src/main/java/dev/langchain4j/model/image/ImageModel.java]: both works, although a bit premature. In any case my main point here is to avoid forcing user to go `.get(0)` all the time

Comment (jdubois) [reply langchain4j-core/src/main/java/dev/langchain4j/model/image/ImageModel.java]: Argh, I couldn't make it work and just read the doc: ` Dall-e-3 models only support a value of 1.`!! So that would only be for DALL E 2, so not really worth it.
I'll just return one value, then.

File: langchain4j-core/src/main/java/dev/langchain4j/data/image/Image.java (hunk 17 above)
Comment (Heezer) [reply langchain4j-core/src/main/java/dev/langchain4j/data/image/Image.java:12]: on the other side other models can simply ignore this field, I opt for this, since it's pity to remove good work done 😄

File: langchain4j-core/src/main/java/dev/langchain4j/data/image/Image.java (hunk 18 above)
Comment (Heezer) [reply langchain4j-core/src/main/java/dev/langchain4j/data/image/Image.java:10]: yes, as discussed in #359, I'm fine with that

File: langchain4j-core/src/main/java/dev/langchain4j/model/image/ImageModel.java (hunk 20 above)
Comment (Heezer) [reply langchain4j-core/src/main/java/dev/langchain4j/model/image/ImageModel.java]: guys, I don't really won't to throw my support of multiple images, instead I would have two methods:
```
public interface ImageModel {
  Response<Image> generate(String prompt);
  default Response<List<Image>> generateMultiple(String prompt) {
    throw new IllegalArgumentException("Operation is not supported");
  }
}
```
like it's non blocking, what do you think?

Comment (langchain4j) [reply langchain4j-core/src/main/java/dev/langchain4j/model/image/ImageModel.java]: Why do you want to have multiple images at once? It is currently supported only with outdated dalle2 model...

Comment (Heezer) [reply langchain4j-core/src/main/java/dev/langchain4j/model/image/ImageModel.java]: If I speak as user, then sure I want to have image variants to my prompt to choose from and technically - the response from DALL·E contains always a list, I suppose DALL·E 3 will support it, this is just matter of time

File: langchain4j-open-ai/src/main/java/dev/langchain4j/model/openai/OpenAiImageModel.java (hunk 21 above)
Comment (Heezer) [reply langchain4j-open-ai/src/main/java/dev/langchain4j/model/openai/OpenAiImageModel.java:89]: the default model is defied in openai4j already, find following code in examples:
````
OpenAiImageModel model = OpenAiImageModel.builder().apiKey(System.getenv("OPENAI_API_KEY")).build();
````

File: langchain4j-open-ai/src/main/java/dev/langchain4j/model/openai/OpenAiImageModel.java (hunk 23 above)
Comment (Heezer) [reply langchain4j-open-ai/src/main/java/dev/langchain4j/model/openai/OpenAiImageModel.java]: intention here was to use the default value defined in openai4j, so if primitive type is used here, then I have to define a default value here again, right?

File: langchain4j-core/src/main/java/dev/langchain4j/model/image/ImageModel.java (hunk 20 above)
Comment (langchain4j) [reply langchain4j-core/src/main/java/dev/langchain4j/model/image/ImageModel.java]: Ok, how about `Response<List<Image>> generate(String prompt, int n)`?

File: langchain4j-open-ai/src/main/java/dev/langchain4j/model/openai/OpenAiImageModel.java (hunk 23 above)
Comment (langchain4j) [reply langchain4j-open-ai/src/main/java/dev/langchain4j/model/openai/OpenAiImageModel.java]: No, no need to define default value for n, it is an optional parameter. So if you have `Integer n` here, and `Integer n` in openai4j, then you can simply do `.n(n)` and not worry about if it is null or not. If it was not befined by user and is null, then it will have no effect and OpenAI's default value will be used. If user defined it, then it will be used.

File: langchain4j-core/src/main/java/dev/langchain4j/model/image/ImageModel.java (hunk 20 above)
Comment (Heezer) [reply langchain4j-core/src/main/java/dev/langchain4j/model/image/ImageModel.java]: sure, I like this!

File: langchain4j-core/src/main/java/dev/langchain4j/data/image/Image.java (hunk 18 above)
Comment (Heezer) [reply langchain4j-core/src/main/java/dev/langchain4j/data/image/Image.java:10]: done, inte test extended

File: langchain4j-core/src/main/java/dev/langchain4j/data/image/Image.java (hunk 19 above)
Comment (Heezer) [reply langchain4j-core/src/main/java/dev/langchain4j/data/image/Image.java]: done

File: langchain4j-core/src/main/java/dev/langchain4j/model/image/ImageModel.java (hunk 20 above)
Comment (Heezer) [reply langchain4j-core/src/main/java/dev/langchain4j/model/image/ImageModel.java]: done

File: langchain4j-open-ai/src/main/java/dev/langchain4j/model/openai/OpenAiImageModel.java (hunk 23 above)
Comment (Heezer) [reply langchain4j-open-ai/src/main/java/dev/langchain4j/model/openai/OpenAiImageModel.java]: this part is obsolete now

File: langchain4j-open-ai/src/main/java/dev/langchain4j/model/openai/OpenAiImageModel.java (hunk 24 above)
Comment (Heezer) [reply langchain4j-open-ai/src/main/java/dev/langchain4j/model/openai/OpenAiImageModel.java:150]: done

File: langchain4j-open-ai/src/test/java/dev/langchain4j/model/openai/OpenAiImagesGenerationIT.java (hunk 25 above)
Comment (Heezer) [reply langchain4j-open-ai/src/test/java/dev/langchain4j/model/openai/OpenAiImagesGenerationIT.java]: done

File: langchain4j-open-ai/src/main/java/dev/langchain4j/model/openai/OpenAiImageModel.java (hunk 26 above)
Comment (Heezer) [reply langchain4j-open-ai/src/main/java/dev/langchain4j/model/openai/OpenAiImageModel.java]: done

**Review by ghost — APPROVED** [review]
@Heezer thank you so much, great job!
